### ISSUE TITLE
Pinpointing lines which cause many tests to fail

### ISSUE THREAD

**Issue by TomNicholas** [coveragepy/coveragepy#785]
### Problem

When you run a large test suite and get many many failures, it's hard to know where the root cause of the problem might be. It would be great if there were some way for your test suite to suggest which functions might be the root of your failures and which are likely just downstream consequences.

### Possible solution

As described [here](https://github.com/pytest-dev/pytest/issues/4906) by @zac-hd, could something like this be done by tracking number of tests failed for each line covered? You would find the lines with the highest ratio of (failing tests covering this line)/(all tests covering this line), and suggest these as the lines to examine first when looking for the reason for the failures.

### Questions

Does something like this already exist? Is there a reason why this couldn't be done?

Is this even the right place to be asking this question? Would this be a feature in `coverage.py`, a plugin for `coverage.py`, an addition to `pytest` or an addition to `pytest-cov`?

**Comment by nedbat** [comment]
This sounds like a really interesting idea.  Coverage 5.0 can now collect data about what tests are covering each line. Perhaps joining that data with something about which tests have passed and failed?

**Comment by TomNicholas** [comment]
Yes if you can access the number of tests which covered this line which failed or passed, you should be able to do it. Does Coverage also know about test success/failure?

**Comment by nedbat** [comment]
I tried an experiment using a junit.xml file to get test/fail, and combining it with the coverage data.  I didn't see how to pinpoint the problem.  Even with one line broken, the results were inconclusive, because some passing tests also ran that line, but didn't assert the wrong outcome the broken line produced.

Here is the code I used:
```
import xml.etree.ElementTree as ET

fails = set()
passes = set()
tree = ET.parse("junit.xml")
for testcase in tree.getroot().findall(".//testcase"):
    testname = testcase.get("classname") + "." + testcase.get("name")
    if testcase.findall("./failure"):
        fails.add(testname)
    else:
        passes.add(testname)
print(f"{len(fails)} failing tests")
print(f"{len(passes)} passing tests")

from coverage import CoverageData

covdata = CoverageData()
covdata.read()
measureds = covdata.measured_files()
print(f"{len(measureds)} measured files")

def run_for_test(testname):
    covdata.set_query_context(testname)
    run = set()
    for measured in measureds:
        lines = covdata.lines(measured)
        run.update(set((measured, line) for line in lines))
    return run

bad_per_file = []
for fail in fails:
    bad = run_for_test(fail)
    if bad:
        bad_per_file.append(bad)

good_per_file = []
for pas in passes:
    good = run_for_test(pas)
    if good:
        good_per_file.append(good)

bads = set.intersection(*bad_per_file)
for g in good_per_file:
    bads -= g
print(len(bads))
```
